### ISSUE TITLE
Review all code for docstrings

### ISSUE THREAD

**Issue by sbailey** [desihub/desispec#17]
All functions should have a docstring with sufficient detail to know what the inputs, outputs, and side effects are.  It shouldn't be necessary to read the code to figure out the units, dimensionality, order of return values etc.  Follow the docstring guidelines from Google:

https://google-styleguide.googlecode.com/svn/trunk/pyguide.html?showone=Comments#Comments

**Comment by dkirkby** [comment]
Did we decide on google vs numpy-style docstrings?  I have been writing google-style comments so far in desispec, but I am willing to convert them if there is a consensus that we prefer numpy (or perhaps the numpy comments in desispec already outnumber the google comments?).

The sphinx config is essentially the same for both styles so this decision has minimal/no impact on installation and packaging.  For reference, here is a brief summary of the differences between the styles: http://sphinx-doc.org/latest/ext/napoleon.html#google-vs-numpy

**Comment by sbailey** [comment]
Let's go with google-style.  I've also been using that (at least approximately).  Thanks for the link to the comparison.

**Comment by weaverba137** [comment]
Before we are too hasty on this...  I did a lot of work to make sure that desiInstall was numpy docstring compatible.  In other words, you can write numpy compatible docstrings in the desiInstall/desiUtil framework without any extra effort.

If you want to just throw that all away, well...

**Comment by weaverba137** [comment]
I'm sorry that was a bit harsh.  Let me rephrase:  Sphinx supports a basic docstring style.  Sphinx can support additional docstring styles through plugins.  The existing desiUtil/desiInstall infrastructure supports the Numpy docstring infrastructure.

So the question is: does Sphinx support Google-style docstrings out of the box, or are additional plugins necessary?  If the latter, it seems to me that there might be some duplication of effort going on.

**Comment by sbailey** [comment]
It appears that the sphinx built-in napoleon extension supports both numpy and google style docstrings:

http://sphinx-doc.org/latest/ext/napoleon.html

The comparison between the two that David linked to was a sub-section of the napoleon docs.  I primarily use the docstrings via pydoc and ipython so I have been immune to formatting syntax differences (though I do care about 80 column width).  If the docstrings aren't formatted perfectly, what happens to sphinx?

**Comment by weaverba137** [comment]
Ah, that looks good, but it does require Sphinx > 1.3, which I don't think we currently have installed at NERSC.

**Comment by weaverba137** [comment]
Hmm, I can't find a sphinx package at NERSC *at all*.  I thought that got installed in January.

**Comment by dkirkby** [comment]
For sphinx < 1.3, the napoleon extension is available as a separately maintained package on pypi:

  https://pypi.python.org/pypi/sphinxcontrib-napoleon/0.3.3

The fact that napoleon has recently migrated into the main sphinx package is encouraging.

**Comment by tskisner** [comment]
Just as another data point, I have been using napoleon and google style doc strings on a new project and am very happy with the result.  See here (click the "[source]" links to see the original doc strings):

http://tskisner.github.io/pytoast/dist.html

packaging sphinx and dependencies (pygments, etc) is on my to-do list for NERSC.  However I still personally feel that documentation is something that should be generated on a laptop, not an HPC system.

**Comment by sbailey** [comment]
All installs should include the documentation.  We shouldn't require users to checkout a copy of the code on their laptops just to build the docs for some version that they are running at NERSC.

**Comment by tskisner** [comment]
If we follow that line of reasoning I think you are headed down a slippery slope.  Are you planning on generating html or latex output with sphinx?  Does that mean that the user needs to run a full texlive installation or a web browser on the NERSC machines?  

I think one has to draw the line somewhere.  Most projects host their online (html) documentation for each tagged version (and maybe dev snapshots).  For example we could host this on github-pages.

Without using sphinx at all, the user has access to the python docs directly in the code.  Claiming that the user must be able to generate sphinx docs and view them all from an HPC system seems really strange to me.

**Comment by tskisner** [comment]
Note that we could set up a github hook to update the docs on readthedocs:

http://docs.readthedocs.org/en/latest/webhooks.html

disclaimer- I have not used this before.

**Comment by weaverba137** [comment]
I *have* used this, & it's not hard, but your documentation builds have to *work*. That is, it doesn't like sloppy documentation that throws lots of Sphinx errors.

It is possible to have Travis builds test documentation builds.

**Comment by dkirkby** [comment]
I have used github hooks to automatically update docs hosted on
readthedocs, most recently here:

  http://weaklensingdeblending.readthedocs.org/

The end result is much more usable for consumers that local docs, so I
think we should aim for this (or something similar).  However, it is a bit
tricky to get a working sphinx config for RTD, so I vote for getting local
doc builds working for desispec first, before we tackle that.

David

**Comment by sbailey** [comment]
A few comments:
* I had been thinking we would use the NERSC website for our documentation, but I like the idea of tying directly to github pages or possibly readthedocs.  This makes getting sphinx working at NERSC less of an issue.  How the docs get built into webpages is a bit off topic for this original ticket, so I spun off a new one for that: ticket #29.

This ticket is about getting the *contents* in place for all classes and functions.
* Since Napoleon supports both numpy and google-style doc strings, let's go with google-style.
* But accurate content with *any* style is better than nothing

**Comment by sbailey** [comment]
Ben did a major pass on cleaning up docstrings and getting sphinx docs working; I did a minor cleanup pass after that.  Docstrings are looking pretty good in most cases.  Closing ticket.

**Comment by sbailey** [comment]
Reopening this ticket.  As @moustakas pointed out, the docstrings are less complete than I had realized.  Several docstrings are simply "No documentation yet." while others are uselessly short (function read_flux_calibration's doctring is "Read flux calibration").  I'm not picky about the formatting, but every function should have a docstring that explains the types and dimensionality of its inputs and outputs.

**Comment by weaverba137** [comment]
Indeed.  My clean up involved getting the docstrings to *work*; I was not concerned about their content.  It is up to the actual authors of the code to supply the content.

**Comment by dkirkby** [comment]
Perhaps a separate docstring ticket for each developer or code feature is a better way to motivate and track this work?

**Comment by sbailey** [comment]
Docstrings updated in PR #64; closing issue (again :)
